File: laboratories/ppo/exercises/lab08/python/src/Collection.py
class Collection:
    def __init__(self, items=[]):
        self.original_items = items
        self.computed_items = items

    def filter(self, function):
        self.computed_items = [item for item in self.computed_items if function(item)]
        return self

    def sort(self, function):
        size = len(self.computed_items) - 1
        for i in range(size):
            for j in range(size - i):
                k = j + 1
                if function(self.computed_items[j], self.computed_items[k]):
                    self.computed_items[j], self.computed_items[k] = self.computed_items[k], self.computed_items[j]
        return self

    def limit(self, limit):
        self.computed_items = self.computed_items[:limit]
        return self

    def get(self):
        return self.computed_items

File: laboratories/ppo/exercises/lab08/php/src/Collection.php
<?php

namespace Ppo\Lab08;

class Collection
{
    protected array $originalItems;
    protected array $computedItems;

    public function __construct(array $items = [])
    {
        $this->originalItems = $items;
        $this->computedItems = $items;
    }

    public function filter(callable $function): static
    {
        $filtered = [];
        foreach ($this->computedItems as $item) {
            if ($function($item)) {
                $filtered[] = $item;
            }
        }

        $this->computedItems = $filtered;

        return $this;
    }

    public function sort(callable $function): static
    {
        $sorted = $this->computedItems;

        $size = count($sorted) - 1;
        for ($i = 0; $i < $size; $i++) {
            for ($j = 0; $j < $size - $i; $j++) {
                $k = $j + 1;
                if ($function($sorted[$j], $sorted[$k])) {
                    list($sorted[$j], $sorted[$k]) = [$sorted[$k], $sorted[$j]];
                }
            }
        }

        $this->computedItems = $sorted;

        return $this;
    }

    public function limit(int $limit): static
    {
        array_splice($this->computedItems, $limit);

        return $this;
    }

    public function get(): array
    {
        return $this->computedItems;
    }
}
